server/hotrod/src/main/java/org/infinispan/server/hotrod/tx/TransactionDecodeContext.java

```diff
@@ -0,0 +1,105 @@
+package org.infinispan.server.hotrod.tx;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.commands.CommandsFactory;
+import org.infinispan.commands.remote.recovery.TxCompletionNotificationCommand;
+import org.infinispan.commands.tx.RollbackCommand;
+import org.infinispan.commons.tx.XidImpl;
+import org.infinispan.commons.util.Util;
+import org.infinispan.factories.ComponentRegistry;
+import org.infinispan.interceptors.locking.ClusteringDependentLogic;
+import org.infinispan.remoting.inboundhandler.DeliverOrder;
+import org.infinispan.remoting.rpc.RpcManager;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.transaction.impl.TransactionTable;
+import org.infinispan.transaction.tm.EmbeddedTransactionManager;
+
+/**
+ * A base decode context to handle prepare, commit and rollback requests from a client.
+ *
+ * @author Pedro Ruivo
+ * @since 9.1
+ */
+abstract class TransactionDecodeContext {
+
+   final XidImpl xid;
+   final ServerTransactionTable serverTransactionTable;
+   final RpcManager rpcManager;
+   final Address localAddress;
+   final TransactionTable transactionTable;
+   final EmbeddedTransactionManager tm;
+   final CommandsFactory commandsFactory;
+   private final ClusteringDependentLogic clusteringDependentLogic;
+   TxState txState;
+
+   TransactionDecodeContext(AdvancedCache<byte[], byte[]> cache, XidImpl xid) {
+      this.xid = xid;
+      this.rpcManager = cache.getRpcManager();
+      this.localAddress = rpcManager != null ? rpcManager.getAddress() : null;
+      ComponentRegistry registry = cache.getComponentRegistry();
+      this.commandsFactory = registry.getCommandsFactory();
+      this.transactionTable = registry.getComponent(TransactionTable.class);
+      this.clusteringDependentLogic = registry.getComponent(ClusteringDependentLogic.class);
+      this.serverTransactionTable = registry.getComponent(ServerTransactionTable.class);
+      this.tm = (EmbeddedTransactionManager) cache.getTransactionManager();
```

Comment:
You don't have to use CacheImpl directly in the server; I suggest overriding `DecoratedCache.readContext`/`writeContext` and pass the transaction identifier to the created context directly.